### DISCUSSION TITLE
How do I get started programming the Pitaya Go?

### DISCUSSION THREAD

**Discussion by LeakyEarth2000** [cmdwtf/Odiin#103] [Q&A]
How do I get started programming the Pitaya Go and to write Chris' code onto it?

**Comment by nitz** [comment]
If you've got a debugging probe/flash tool that is supported by [`pyOCD`](https://pyocd.io/) (Most CMSIS-DAP probes should work, I use the [Pitaya-Link](https://makerdiary.com/products/pitaya-link)), you should be good to go.

You'll want to ensure you've got your pins connected to the Pitaya go on at least the first 5 pins of the right header (the side opposite the battery connector.) That would be the `3.3V` pin (power supply), `GND`, `/RESET`, `SWCLK` and `SWDIO`. I've attached the pitaya go pinout to the end of this post for your reference!

You want to start flash the NRF's MBR first (a tiny binary blob provided by Nordic that supports the bootloader, etc.) You only need to flash it once, which you can do with `make flash_mbr`, or by using the command from the `Makefile` directly: [here](https://github.com/cmdwtf/Odiin/blob/main/Makefile#L553).

Then, it should be as simple as running `make flash_all`. That would build the bootloader, and main application firmware and flash it to the board!

A few references: 
- The [tasks.json](https://github.com/cmdwtf/Odiin/blob/main/.vscode/tasks.json) that VSCode references for commands I use regularly when developing, mostly just shortcuts to different `make` targets.
- The main [targets in `Makefile`](https://github.com/cmdwtf/Odiin/blob/main/Makefile#L541-L582) referenced from the tasks.json, but to give you an idea of what to run if you wanted to run `pyocd` and the sort to flash manually without using `make`.


A side note:
If you wanted to use the 'end user' method of updating the application, you could do `make flash_bootloader`, which will just build and flash the bootloader portion, which would enable you to use the UF2 bootloader to update the application on the Pitaya Go via USB. If you choose to go this route of manually flashing the application separate from the bootloader, you'll need to poke a `1` value into a specific spot in the flash to tell the bootloader that the application is valid and it should try to run it. That is handled in the `Makefile` automatically [here](https://github.com/cmdwtf/Odiin/blob/main/Makefile#L571), but you can manually do so via: `pyocd cmd -t nrf52840 --command write32 0xFF000 0x00000001 --command reset`. Without this value set, the bootloader won't ever pass execution to the application.

Let me know if you run into any issues!

Pitaya Go pinout:

![pitaya go pinout](https://user-images.githubusercontent.com/602691/203558424-1a7fe1f6-3b94-4a5d-aef5-04c0a7cc5261.png)

**Reply by LeakyEarth2000** [reply]
> You want to start flash the NRF's MBR first (a tiny binary blob provided by Nordic that supports the bootloader, etc.) You only need to flash it once, which you can do with `make flash_mbr`, or by using the command from the `Makefile` directly: [here](https://github.com/cmdwtf/Odiin/blob/main/Makefile#L553).

I'm a little confused here because I actually don't know what software to use to program this line of code. Even on the [wiki](https://wiki.makerdiary.com/pitaya-go/programming/), where it shows you how to program using the CMSIS, if I copy the command onto CMD with Admin, it just doesn't work. I have installed the other pip libraries. But, I have no idea how to go forward.

**Reply by nitz** [reply]
Can you run `pyocd` that you installed via `pip` or as in the instructions [here](https://wiki.makerdiary.com/pitaya-go/programming/#using-a-cmsis-dap-debug-probe)? If you can, try running `pyocd erase -t nrf52840 --chip`, which will do a full wipe of the nRF52840, and confirm that your link is at the very least working!

**Reply by LeakyEarth2000** [reply]
`'pyocd' is not recognized as an internal or external command,
operable program or batch file.` is what comes up if I do any command, despite installing the pyocd library. I did both `pip install pyocd`, `pip install -U pyocd`, both of which tell me the requirements are satisfied, so already been installed; and also `pip install --pre -U https://github.com/mbedmicro/pyOCD/archive/master.zip`.

**Reply by LeakyEarth2000** [reply]
What if I ditch using PyOCD and just use VSC? Could that still work?

**Reply by nitz** [reply]
Visual Studio Code just has hooks that call into `make` that eventually will call `pyocd`. You can probably do the majority of it manually with `nrfutil`, but I'm not sure what all it's capable of.

I'm thinking the reason you can't find `pyocd` is just a system path issue. Find out where pip put the "script" executable for `pyocd` (You'll be looking for `pyocd.exe`. If you run this one-liner from the command prompt, it should give you a good idea as to which directories you might want to check in: `python -c "import sys; print('\n'.join(sys.path))"`)

Then you'll want to add that directory to your path (which you can [like so](https://www.computerhope.com/issues/ch000549.htm)), which will make it easier to call and also work from the make scripts. Alternatively you can just manually use the full path to the exe every time you want to use it, e.g.:

```
"c:\my\path to python\Scripts\pyocd.exe" erase -t nrf52840 --chip
```

**Reply by LeakyEarth2000** [reply]
Yes! It finally works when I do `pyocd erase -t nrf52840 --chip`

**Reply by LeakyEarth2000** [reply]
Two things:

I'm getting this message `Board ID 1028 is not recognized [mbed_board]` every time I run any command. Is this fine?

Also, when I run the command line `pyocd flash -t nrf52840 $(SDK_ROOT)/components/softdevice/mbr/hex/mbr_nrf52_2.4.1_mbr.hex`, I get the error `Error: [Errno 2] No such file: '$(SDK_ROOT)\components\softdevice\mbr\hex\mbr_nrf52_2.4.1_mbr.hex' [__main__]`. I understand I have to get some file, but am I able to know what file specifically? And where to save it at?

> Then, it should be as simple as running `make flash_all`. That would build the bootloader, and main application firmware and flash it to the board!
I'm also confused here. Another file I have to run?

If it isn't obvious here, I'm new to the Pitaya and am a complete beginner in the Pitaya family :)

**Reply by nitz** [reply]
> I'm getting this message `Board ID 1028 is not recognized [mbed_board]` every time I run any command. Is this fine?

Probably! I'm not quite sure where it's getting that `1028` from, but, I'm guessing it's likely something being provided as an ID of some sort that just doesn't match things in `pyocd`'s database. (I'm using an older version of `pyocd`, but if I do run across that message I'll let you know.)

> I understand I have to get some file, but am I able to know what file specifically? And where to save it at?

So the file you're looking for is `mbr_nrf52_2.4.1_mbr.hex`, and it's part of the nRF 5 SDK, which is in a git submodule of the repo. If you cloned the repo with `--recursive`, then you should already have it locally. If you didn't, run `git submodule update --init` from the repo root and it'll fetch all the submodules. Then you just need to replace `$(SDK_ROOT)` with the path to where that submodule is, 

For example: `C:\your\path\to\Odiin\vendor\nRF5_SDK\components\softdevice\mbr\hex\mbr_nrf52_2.4.1_mbr.hex`

> If it isn't obvious here, I'm new to the Pitaya and am a complete beginner in the Pitaya family :)

No worries, there's a lot of moving parts here. `make` is [GNU Make](https://www.gnu.org/software/make/), which is the application that processes `Makefiles` (like [Odiin's](https://github.com/cmdwtf/Odiin/blob/main/Makefile),) to run a sequence of other applications (compilers, linkers, flash utilities, etc.) When you run `make flash_all` from the repository root, `make` will parse the `Makefile`, look for and execute the commands for the recipe titled `flash_all`, which is here on line [568](https://github.com/cmdwtf/Odiin/blob/main/Makefile#L568). Targets are the lines that begin with `some_name:`, and any of the lines that are spaced over under it are commands that are executed after all of the required targets (on the same line as the name) finish. `flash_all`, runs the `merge` target, which runs the `bootloader` and `default` targets before merging the two outputs of those into a single hex file. That hex file is what the `flash_all` target then attempts to flash to the pitaya with the `pyocd` command.

If you've got all the tools from the "[Build Enviornment](https://github.com/cmdwtf/Odiin#build-environment)" section installed, and have configured the paths in `[BuildSettings.mk](https://github.com/cmdwtf/Odiin/blob/main/BuildSettings.mk)`, then running `make flash_all` should go through each step of the compilation, generate the merged hex file, and attempt to flash it.

**Reply by LeakyEarth2000** [reply]
I have almost completed everything. Just need to do the GNU Makefile part.

So I have installed the nRF5 command line tools and the ARM GNU Toolchain. But, I'm unsure what to do after. I am aware of a CMD prompt opening from ARM GNU Toolchain and my best guess is having to use this somehow.

[Lines 569 - 571](https://github.com/cmdwtf/Odiin/blob/main/Makefile#L569-L571) exclaims `(OUTPUT DIRECTORY)` and `(FILENAME_OUTPUT_MERGED_HEX)`. Exactly, what would the output directory be and is the merged hex the same mbr.hex file? How would the code in the command line be then for these lines? Do I have to do every line individually on the cmd or all together? Are these to be in the commanded in a standard CMD admin or to open up in ARM GNU Toolchain file location?

I don't want to accidentally screw anything up, and just would like to be sure.

**Reply by LeakyEarth2000** [reply]
![IMG20221205141812.jpg](https://user-images.githubusercontent.com/117555864/205555442-8cbae986-cf37-4665-bbcc-6fe3eb371720.jpg)
![IMG20221205161832.jpg](https://user-images.githubusercontent.com/117555864/205555408-3016b2f4-7f51-43bf-8d0a-86ddbaba3d5e.jpg)

A hardware question too. Is it best to solder these female headers to the bridge or should I directly solder the LCD to it? It's currently being blocked by the SD card to be flat, so what do you propose? Should I just raise it half way then solder the pins? Like here: ![IMG20221205141854.jpg](https://user-images.githubusercontent.com/117555864/205555605-daae76e6-9ed0-448b-958a-07269551437d.jpg)

**Reply by nitz** [reply]
Oh goodness, apologies I missed your earlier reply!

> So I have installed the nRF5 command line tools and the ARM GNU Toolchain. But, I'm unsure what to do after. I am aware of a CMD prompt opening from ARM GNU Toolchain and my best guess is having to use this somehow.

Now you just need to edit the `BuildSettings.mk` file to have the correct directories to where your toolchain is installed.

Then, either from the command line while in the Odiin directory, you can run `make flash_all`. Or, if you've got the workspace open in VS Code, you can press <kbd>CTRL</kbd>+<kbd>SHIFT</kbd>+<kbd>B</kbd> to run the command that way.

> Exactly, what would the output directory be and is the merged hex the same mbr.hex file?

Eventually it spits the resulting files out in a directory named `_build`. The final merged hex file will be named something like `odiin-[git hash].hex` (and `odiin-[git hash].uf2` for flashing via the UF2 bootloader.)

> How would the code in the command line be then for these lines? Do I have to do every line individually on the cmd or all together?

When you run `make flash_all`, it will run all of those in sequence. (After it finishes the `merge` target.)

> Are these to be in the commanded in a standard CMD admin or to open up in ARM GNU Toolchain file location?

You shouldn't need to run as admin. As long as your `GNU_INSTALL_ROOT` and `GNU_VERSION` is set properly in `BuildSettings.mk`, it will invoke calls to the toolchain directly. The things you'll need on your path are `make` and `pyocd`. Perhaps some others I'm forgetting, but I know at least those two.

> I don't want to accidentally screw anything up, and just would like to be sure.

Good news is that you basically can't screw anything up. Software can always be rebuilt and if the wrong thing gets flashed to the board, wiping it and flashing again is no problem.

> A hardware question too. Is it best to solder these female headers to the bridge or should I directly solder the LCD to it? It's currently being blocked by the SD card to be flat, so what do you propose? Should I just raise it half way then solder the pins? Like here: 

Headers are absolutely your call. I soldered the LCD directly to the bridge because I was trying to minimize height to fit in a printed case. As long as the SD Card slot isn't touching the solder joints below it, you're fine.

**Reply by LeakyEarth2000** [reply]
> Oh goodness, apologies I missed your earlier reply!

Nah, no worries! Take your time to reply! We're in no rush :)

> Then, either from the command line while in the Odiin directory, you can run `make flash_all`. Or, if you've got the workspace open in VS Code, you can press CTRL+SHIFT+B to run the command that way.

Sorry, I'm a little bit confused about `make flash_all`. Could you please send the whole command to type and execute whilst in the Odiin directory? That would be greatly appreciated. I would just CTRL + SHIFT + B in VSC, but when I do, it just says this and I really can't be bothered going back and forth to fix this.

```
make : The term 'make' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and 
try again.
At line:1 char:1
+ make release
+ ~~~~
    + CategoryInfo          : ObjectNotFound: (make:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

> Eventually it spits the resulting files out in a directory named _build. The final merged hex file will be named something like odiin-[git hash].hex (and odiin-[git hash].uf2 for flashing via the UF2 bootloader.)

With the hex file, what should I do with it? And with the uf2 file, what should I do afterwards with it, too? It would be amazing if you could list the commands to execute 1-by-1. 

I also massively apologise for being such a burden

**Reply by LeakyEarth2000** [reply]
I understand you may be busy, but I was hoping you could take a moment to review this message when you have a chance. Thank you, Chris!

**Reply by nitz** [reply]
Apologies I did get busy with the holidays and my day job 😅

> I also massively apologise for being such a burden

Not at all, we all get to learn new things every day!

> Sorry, I'm a little bit confused about `make flash_all`. Could you please send the whole command to type and execute whilst in the Odiin directory? That would be greatly appreciated. I would just CTRL + SHIFT + B in VSC, but when I do, it just says this and I really can't be bothered going back and forth to fix this.

`make flash_all` is the whole command :) It runs the `make` utility with the `flash_all` target, which ends up running the whole slew of commands to get you there.

The error you're getting (`'make' is not recognized...) just means you don't have the `make` tool installed (or available via your system's PATH environment variable.) Try installing a pre-built copy of it from here: https://gnuwin32.sourceforge.net/downlinks/make.php, and then running the command. (You'll need to re-start your terminal or vscode as applications only see changes to environment variables on launch.)

`make flash_all` from the terminal (or <kbd>CTRL</kbd>+<kbd>SHIFT</kbd>+<kbd>B</kbd> in vscode should run all the commands you need. If you want to see what commands `make` would execute, run from the terminal: `make --dry-run flash_all` and it will show you what make runs for that target, like such:

```
Y:\workspace\Odiin>make --dry-run flash_all
echo DONE nrf52840_xxaa
echo Merging: _build/odiin-a2fcd1d.hex from application and UF2 bootloader.
mergehex -m _build/uf2_bootloader.hex _build/nrf52840_xxaa.hex -o _build/odiin-a2fcd1d.hex
echo Flashing: _build/odiin-a2fcd1d.hex
pyocd flash -t nrf52840 _build/odiin-a2fcd1d.hex
pyocd cmd -t nrf52840 --command write32 0xFF000 0x00000001 --command reset

Y:\workspace\Odiin>:: notice that --dry-run only shows the commands for the requested target in it's current state, so your output will be considerably longer as it will do all the targets it depends on, too
```

Let me know how you get on with that, and then we can turn back to the hex/uf2 file stuff.

**Comment by LeakyEarth2000** [comment]
Hey Chris! It's a been a very long time, but I am finally back to continue this project!

I have no idea what step I am up to, so I think it's best for me to reset my Pitaya Go. Problem is that none of the reset/user buttons actually work no longer. If I hold the user or reset button before plugging the power in, nothing happens! The green LED lights up, but no input. With my pitaya link, if I hold the only button (boot/reset) before plugging the power in directly to it, it will pop up a drive on my PC. with a firmware file. I tried putting it in DFU, but nothing happens. In fact, even if I plug in the pitaya go to my PC, my PC doesn't make that hardware connect sound and device manager doesn't acknowledge the device at all

As I said, I think it is best that I reset my Pitaya Go, but I have no idea how to now. I thought it would have been simple as holding the reset button down and plugging the USB C cable in! If you could guide me on what was the software needed in order to perform the reset, as well as the command to actually reset this thing, that would be greatly appreciated!

Cheers
